### ISSUE TITLE
decoding 'null' for a case class with only Option fields does not fail fast as expected

### ISSUE THREAD

**Issue by longcao** [travisbrown/circe#217]
@ty707 and I found a bug with automatic decoding + derivation in which when decoding a `null` to a case class of only `Option` fields, it doesn't fail fast as expected - it succeeds and returns an instance of the case class where all of its fields are `None`.

Example error was repro'd on master @ dee431a5cf364f4ed4cd113f494bfe84cc826f11.

Pasting my REPL session:

```scala
import io.circe._
import io.circe.generic.auto._
import io.circe.literal._
import io.circe.parser._
import io.circe.syntax._
import cats.data.Xor

scala> case class Container[T](data: T)
defined class Container

scala> case class Item(a: Option[Int], b: Option[String])
defined class Item

// I'd expect this to fail hard because data: T was improperly defined, since the "data" field is in fact nullable/optional
scala> decode[Container[Item]]("""{"data":null}""")
res0: cats.data.Xor[io.circe.Error,Container[Item]] = Right(Container(Item(None,None)))

// I'd also expect this to fail hard
scala> decode[Item]("""null""")
res1: cats.data.Xor[io.circe.Error,Item] = Right(Item(None,None))

scala> case class Qwer(a: Option[Int], b: String)
defined class Qwer

// This seems to behave as expected, having a single non-Option field causes it to fail hard
scala> decode[Qwer]("""null""")
res2: cats.data.Xor[io.circe.Error,Qwer] = Left(io.circe.DecodingFailure: Attempt to decode value on failed cursor: El(DownField(b),false))
```

**Comment by travisbrown** [comment]
This is absolutely a bug (inherited from Argonaut). I've posted a quick draft of a fix in #218 and will wrap it up and backport it to 0.3 as soon as possible.

Thanks for reporting, @longcao and @ty707—hope it didn't cause too many problems for you.

**Comment by longcao** [comment]
@travisbrown No problem, this was really an unlucky combination of both user error and `circe` not failing hard enough - in our case we were incorrectly assuming our `data: T` field was to always return data when it reality it should have been correctly modeled as `data: Option[T]`. The expectation would be that `circe` would fall down trying to do this at runtime.

**Comment by travisbrown** [comment]
Fixed in #231.
